### ISSUE TITLE
tab groups / tagging of tabs

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#49]
For people with many tabs, it might make sense if they'd be able to tag/label them and filter them.

**Comment by bb010g** [comment]
:+1:

**Comment by vyp** [comment]
I haven't actually got the time to test out qutebrowser yet, so apologies if this already happens. But maybe also an option that if enabled, automatically groups by domain too. This isn't exactly the same thing as tags/labels/groups of course, but related enough I thought to put the idea here.

**Comment by The-Compiler** [comment]
I like those ideas, but don't expect me to work on them anytime soon - I've got enough on my plate with some big things already (see the bottom of [this mail](https://lists.schokokeks.org/pipermail/qutebrowser/2015-August/000105.html) for details).

Of course (as always), contributions would be very appreciated!

**Comment by vyp** [comment]
@The-Compiler No worries of course, thanks for letting know. Just putting it out there.

**Comment by lamarpavel** [comment]
@vyp, why not automatically add a tag for every base URL to the tabs, that way when grouping by tags is enabled, tabs will automagically be grouped by URL. Unless I'm missing something this would avoid two differnt solutions for a very similar feature and indeed make use of an already implemented system.

**Comment by vyp** [comment]
@lamarpavel Good idea. I guess the automatic tags could be the domain name? (e.g. "github" for github.com)

**Comment by lamarpavel** [comment]
@vyp I said base URL because it's unique, some websites have the same domain name with different TLDs, eg amazon.com and amazon.co.uk and sometimes even with completely different sites. Though I guess that wouldn't be much off an issue because ist's rare and wouldn't cause much harm anyway, so domain name without TLD might be better.

**Comment by vyp** [comment]
@lamarpavel Actually no I think combining the tld might be better, for what you just said. Same domain names on different TLDs can still be vastly different. One example that comes to mind is gnu.org vs gnu.com. So clearly grouping them the same would not be right. I think there's no harm in putting the tld in the tag too, what do you think? (Is that what you mean by the base URL?)

**Comment by lamarpavel** [comment]
Sorry for the late reply, @vyp; by base-url I meant the domain name plus the tld, yes.

**Comment by vyp** [comment]
Cool thanks. :+1:

**Comment by bzar** [comment]
I'd envision a minimal implementation of this feature working with a "current-tags" state.

- Setting "current-tags" filters the tabs to only those with all of the tabs in the list.
- Additionally, any new tabs are automatically tagged with "current-tags".
- By default "current-tags" is empty and doesn't affect users in any way.
- If "current-tags" is set, its value is shown somewhere in the tab list.

**Comment by proxict** [comment]
Sorry to necrobump this issue, but I've seen other people opening new issues with the same topic and this issue is usually linked there.

I've been using and loving qutebrowser since 2017, and I'm a little embarrassed to admit that I still haven't figured this out:
I'm a massive tab abuser; it's very common for me to have over 70 tabs open. Some of them, I keep open for over a year and only then I consider them as candidates for closing.
Due to this bad habit of mine, it often becomes difficult for me to navigate between all the tabs when I am dove into a specific topic and I have loads of additional tabs related to that topic open for several days.
I have tried to solve this using sessions (`:session-load`, `:session-save`, `:session-delete`), however, that's very impractical for frequent context switching as all the tabs get reloaded when hopping between sessions.

What I would love would be to have multiple tab-groups that I could toggle on the fly.
Something like:
 - `:tab-group-new "rust tokio"`: creates a new tab-group called "rust tokio", hides all tabs from "default" tab-group and displays "about:blank" (or default home-page).
 - `:tab-group-select default`: hides all tabs from the "rust tokio" tab-group and shows all tabs from "default" tab-group.
 - `:tab-group-delete "rust tokio"`: closes all tabs from "rust tokio" and switches to "default" tab-group.
 - `:tab-group-new --temp "cat pics"`: creates a new "cat pics" tab-group that isn't persistent across qutebrowser restarts (`--temp`).

Under the hood, all the tabs in all tab-groups would be loaded to allow for fast context switching, but only one tab-group would be visible at a time. Or perhaps it could be optional to have all tab-groups visible below each other as separate tab-bars.
Either way, I believe this would be a killer feature!

Or maybe, though I've looked everywhere I could think of, there already is such a mechanism that allows for the workflow I've described above and I'm going to be even more embarrassed not to know about it.
